TS: Add generic types for Geometry and Material definitions

jedahan opened the pull request:
This allows for inferences like .getAttribute() on a THREE.Points .geometry instance, which is nice. I also ran eslint --fix, which for some reason changed all the imports from `'` to `"`.

Mugen87 commented:
> I also ran eslint --fix, which for some reason changed all the imports from ' to ".

Please try to undo this changes since this makes it hard to see the effective changes of your PR.

jedahan commented:
@Mugen87 I split it into two commits so you can just view the relevant one https://github.com/mrdoob/three.js/pull/18720/commits/3145ec8cb2e877d53037f4bef7777b2b14af2acb . Might need to fix eslint config separately.

mrdoob commented:
> @Mugen87 I split it into two commits so you can just view the relevant one 3145ec8 . 

Better to separate it in different PRs 🙏

Commit:
Use generic geometry and material typescript types

This allows for inferences like .getAttribute() on a
THREE.Points .geometry instance.

jedahan commented on src/objects/InstancedMesh.d.ts:
```diff
@@ -1,17 +1,16 @@
 
-	constructor(
-		geometry: Geometry | BufferGeometry,
-		material: Material | Material[],
```
The docs say it takes a Material https://threejs.org/docs/index.html#api/en/objects/InstancedMesh

mrdoob replied:
Correct. I don't think it supports `Material[]`.

jedahan commented:
I am new to typescript, and three.js so if anything looks slightly out of whack don’t hesitate to ask questions or point it out, I’m sure some of it is incorrect or broken.

mrdoob commented:
I don't have experience with TypeScript.
Any chance you can find someone that can double check your change?

mrdoob commented:
/ping @FMS-Cat

FMS-Cat commented:
Oh hi the GitHub Mobile notification, you're the best

FMS-Cat commented on src/objects/Mesh.d.ts:
```diff
@@ -1,19 +1,20 @@
-export class Mesh extends Object3D {
+export class Mesh<
+	TGeometry extends Geometry | BufferGeometry = BufferGeometry,
+	TMaterial extends Material | Material[] = MeshBasicMaterial
```
Why did you set the default type to `MeshBasicMaterial` instead of `Material | Material[]` ?
It will fail on most of existing TypeScript codes when we use `Mesh` for interfaces or types.

I assume you were intended to indicate the type of default constructors, and it works fine when we just do `new Mesh()` .
But we have to replace every existing `Mesh` in interfaces with `Mesh<THREE.Geometry | THREE.BufferGeometry, THREE.Material | THREE.Material[]>` (or something strict) 😩 
I'm not sure whether this is an acceptable change or not.

FMS-Cat replied:
@Mugen87 how do you think about it?

Mugen87 replied:
Sorry, I'm not into TS enough to review this code.

petamoriken replied:
How about this type declarations?

```ts
interface Mesh<
    TGeometry extends Geometry | BufferGeometry = Geometry | BufferGeometry,
    TMaterial extends Material | Material[] = Material | Material[],
> extends Object3D {
    geometry: TGeometry;
    material: TMaterial;

    // ...
}

declare var Mesh: {
    prototype: Mesh;
    new<
        TGeometry extends Geometry | BufferGeometry = BufferGeometry,
        TMaterial extends Material | Material[] = MeshBasicMaterial,
    >(geometry?: TGeometry, material?: TMaterial): Mesh<TGeometry, TMaterial>;
}
```

FMS-Cat replied:
My coworker gave me a great suggestion!! He says he has referenced one of dom.d.ts declaration.

FMS-Cat replied:
I just have confirmed this works pretty well on my existing codes. @jedahan could you check this out

jedahan replied:
Sorry, been busy. Will likely be able to get to this tomorrow evening.

FMS-Cat commented:
@jedahan I just looked through the changes, I got the point, and I love it.
There is one point I couldn't judge by myself (commented above).

FMS-Cat commented on src/objects/SkinnedMesh.d.ts:
```diff
@@ -1,15 +1,19 @@
-export class SkinnedMesh extends Mesh {
+export class SkinnedMesh<
+	TGeometry extends BufferGeometry,
+	TMaterial extends Material = MeshBasicMaterial
```
```suggestion
	TMaterial extends Material | Material[] = MeshBasicMaterial
```

pailhead replied:
@FMS-Cat why is it needed to have the assignment here? Not asking about the basic material in particular but also having `= Material | Material[]`?

FMS-Cat replied:
see the discussion: https://github.com/mrdoob/three.js/pull/18720/files#r406096587

pailhead replied:
yeah no figured it out, i think the other PR is the same as this now

FMS-Cat commented on src/objects/SkinnedMesh.d.ts:
```diff
@@ -1,15 +1,19 @@
 
-export class SkinnedMesh extends Mesh {
+export class SkinnedMesh<
+	TGeometry extends BufferGeometry,
```
I prefer having `= BufferGeometry` here, depends on how [the discussion](https://github.com/mrdoob/three.js/pull/18720/files#r406096587) goes

Mugen87 commented:
Should be fixed via #19299.
